extensions/cornerstone/src/Viewport/OHIFCornerstoneViewport.tsx

```diff
@@ -228,12 +228,13 @@ const OHIFCornerstoneViewport = React.memo(props => {
 
       CornerstoneViewportService.setViewportDisplaySets(
         viewportIndex,
-        viewportData,
+        viewportDataLoaded,
         viewportOptions,
         displaySetOptions
       );
 
-      setViewportData(viewportData);
+      setViewportData(viewportDataLoaded);
+      CornerstoneCacheService.setViewportData(viewportDataLoaded, viewportIndex);
```

Comment:
Explicitly call the cache service viewport data changed event, so that it happens after the actual change has completed.

Reply:
This doesn't make sense to me, the CacheService is returning the data itself, and we are setting it here again just for the purpose of event firing. I guess this is a fix that you have for the race condition you are seeing but will introduce more bugs. `setViewportData` is an async operation by react for state update. I guess the event was firing before this react state is set (I'm not even sure we need to keep track of the viewportData here at all, but have not had the chance to debug it)

Reply:
Not everyone has access to "viewportData" in the react sense, but still needs to know when the viewport data has been loaded, so that they can rely on viewport specific things occurring.  The problem is, most users of the data use it by accessing the cornerstone3D viewport, which has multiple events which fire, so there isn't a good way of knowing which one to use.  This event now becomes a reliable "viewport is loaded" event.  I'm not totally happy with it yet, because it isn't cross viewport types - eg an ECG viewport should also be able to fire it, but currently can't.